grpclb/src/main/java/io/grpc/grpclb/CachedSubchannelPool.java

```diff
@@ -38,28 +39,39 @@
   private final HashMap<EquivalentAddressGroup, CacheEntry> cache =
       new HashMap<>();
 
-  private Helper helper;
-  private LoadBalancer lb;
+  private final Helper helper;
+  private PooledSubchannelStateListener listener;
 
   @VisibleForTesting
   static final long SHUTDOWN_TIMEOUT_MS = 10000;
 
-  @Override
-  public void init(Helper helper, LoadBalancer lb) {
+  public CachedSubchannelPool(Helper helper) {
     this.helper = checkNotNull(helper, "helper");
-    this.lb = checkNotNull(lb, "lb");
   }
 
   @Override
-  @SuppressWarnings("deprecation")
+  public void registerListener(PooledSubchannelStateListener pooledSubchannelStateListener) {
+    this.listener = checkNotNull(pooledSubchannelStateListener, "pooledSubchannelStateListener");
+  }
+
+  @Override
   public Subchannel takeOrCreateSubchannel(
       EquivalentAddressGroup eag, Attributes defaultAttributes) {
     final CacheEntry entry = cache.remove(eag);
     final Subchannel subchannel;
     if (entry == null) {
-      // TODO(zhangkun83): remove the deprecation suppression on this method once migrated to the
-      // new createSubchannel().
-      subchannel = helper.createSubchannel(eag, defaultAttributes);
+      subchannel =
+          helper.createSubchannel(
+              CreateSubchannelArgs.newBuilder()
+                  .setAddresses(eag)
+                  .setAttributes(defaultAttributes)
+                  .build());
+      subchannel.start(new SubchannelStateListener() {
+        @Override
+        public void onSubchannelState(ConnectivityStateInfo newState) {
+          listener.onSubchannelState(subchannel, newState);
```

Comment:
It seems you make `CachedSubchannelPool#handleSubchannelState()` private  and call  `handleSubchannelState()` here directly. Then you don't need the API `SubchannelPool#handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo newStateInfo)` in the interface anymore.  Otherwise, `PooledSubchannelStateListener#onSubchannelState()` calls `GrpclbState#handleSubchannelState`, which calls `CachedSubchannelPool#handleSubchannelState` back and forth.

Reply:
i am little bit confused. first method mean `Grpclb#handleSubchannelState(Subchannel, ConnectivityStateInfo)`?

Reply:
I mean you can make `CachedSubchannelPool#handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo newStateInfo)` a private method. Remove this method from interface. And call this private method directly here.

```java
listener.onSubchannelState(subchannel, newState);
handleSubchannelState(subchannel, newState);
```

Do not call this private method from the other class `GrpclbState`.

Reply:
oh yes. done PTAL

Reply:
Then we don't even need the private method, we can just inline it here
```java
CacheEntry cached = cache.get(subchannel.getAddresses());
if (cached != null && cached.subchannel != subchannel) {
  cached.state = newState;
}
```

Reply:
i don't see why having a small method is bad. even though it is short, method name is helpful to read the code.

Reply:
okay. I believe smaller method is almost always preferred, so problem here is the method name. how about `updateCachedSubchannelState`?